File: ruby/lib/email_validator/parser.rb
require 'parslet'

class EmailValidator::Parser < Parslet::Parser
  rule(:local_part) { match['a-zA-Z'].repeat(1) }
  rule(:domain_char) { match['^@\\.'] }
  rule(:at)         { str('@') }
  rule(:dot)    { str('.') }
  rule(:sub_domain) { domain_char.repeat(1) }
  rule(:domain) { (sub_domain >> dot).repeat(1) >>
                  sub_domain }
  rule(:email)  { local_part >> at >> domain }
end

File: python/email_validator/parser.py
from pyparsing import *

ParserElement.setDefaultWhitespaceChars('')

local_part = OneOrMore(Regex(r'[a-zA-Z]'))
domain_char = Regex(r'[^@\.]')
at = Literal('@')
dot = Literal('.')
sub_domain = OneOrMore(domain_char)
domain = OneOrMore(sub_domain + dot) + sub_domain
email = local_part + at + domain

def parse(s):
    return email.parseString(s)
